docs/adr/0001-add-ons-import.md

```diff
@@ -0,0 +1,41 @@
+# 1. Add ONS Postcode Data Import
+
+Date: 2023-05-09
+
+## Status
+
+Open
+
+## Context
+
+Locations API is supposed to be an exhaustive list of postcodes, but because the underlying API (OS Places API) is a list of addresses, postcodes which aren't the primary postcode for an address in the OS Places API database won't ever appear. So Locations API can't support Large User Postcodes (LUPs), or Retired Postcodes.
+
+The nature of data in Imminence means that a lot of LUPs are provided as the contact information for places in Imminence, and without LUPs in the Locations API we can neither geocode not search by these postcodes. The current workaround is to manually look up the Longitude and Latitude of the postcode in the Office of National Statistics Postcode Data (ONSPD) file, which contains information as a CSV list of postcodes, and critically includes both LUPs and retired postcodes. This is not possible for content second liners, and not easy for technical second liners, so second line tickets generated by these problems are hard to resolve without toil work on the part of a handful of specialists.
+
+It also means we cannot support end users who only have retired postcodes for their addresses (more common than expected, because it seems the communications around postcode retirement are often missed).
+
+## Decision
+
+Since the ONSPD file is updated relatively frequently but not too frequently, and has a well understood structure, we will  allow for it to be imported into the postcode table in Locations API.
+
+To differentiate these import records from OS Places API records, two fields will be added to the database
+- a source field, an enum which will record where the data comes from, and
+- an active field, a boolean field which will note whether the postcode is active or retired.
+
+The rest of the ONSPD data for a postcode will be stored in a similar JSON structure to that we currently use for OS Places API-derived data, but with a simplified structure.
+
+We will create import workers to handle the import. A rake task will be created which, given the URL of an ONSPD zip file, will schedule a download worker task. The download worker will retrieve the zip file and extract the smaller CSV files, putting them in an S3 bucket. It will then schedule import worker tasks for each CSV file. These tasks will read the CSV, comparing it with the existing database. Where a postcode apppears in the CSV but is not in the current database, it will create a simplified record for it, and mark down the source as ONSPD. These workers will not override any record that has been retrieved from the OS Places API.
```

Comment:
I think given that 
- it's rare (2-3 times a year) 
- it's not necessarily predictable when a new file comes out, and
- the process of getting the link is easy for a human but awkward for a machine (it currently involves going to the ONS site, selecting something from a dropdown menu, eyeballing that this is the latest, grabbing the link)

...to start with this will just be a manually called job rather than putting in some brittle code to grab the URL based on the current structure of their website. Basically, we'll put it in a diary for someone to check every 6 months.

I emailed them to ask if there's a permalink that always points to the most recent file (in which case this could be automated more safely), but haven't heard back.

Reply:
Ok, so human gets the file and uploads it to s3, and rake task points to url of file in the bucket? 👍 .

Reply:
It's more like:
- human gets the URL and passes that to the rake task
- rake task downloads the file and puts bits of it in in the S3 bucket
- (then secondary tasks take the bits and process them).

This way human just has to go to a website and copy a link, then run a rake task with that link. Human doesn't have to faff around with downloading the file, logging into AWS, then finding the right bucket to put the file in.